File: src/indexing/embedding.js
const { normalizeText } = require('../utils/text');

function fnv1aHash(str) {
  let hash = 2166136261;
  for (let i = 0; i < str.length; i += 1) {
    hash ^= str.charCodeAt(i);
    hash = (hash * 16777619) >>> 0;
  }
  return hash >>> 0;
}

function embedText(text, options) {
  const dims = options.dims || 384;
  const ngramMin = options.ngramMin || 3;
  const ngramMax = options.ngramMax || 5;

  const normalized = normalizeText(text);
  const padded = ` ${normalized} `;
  const vector = new Array(dims).fill(0);

  for (let n = ngramMin; n <= ngramMax; n += 1) {
    for (let i = 0; i <= padded.length - n; i += 1) {
      const gram = padded.slice(i, i + n);
      const idx = fnv1aHash(gram) % dims;
      vector[idx] += 1;
    }
  }

  let norm = 0;
  for (let i = 0; i < vector.length; i += 1) {
    norm += vector[i] * vector[i];
  }
  norm = Math.sqrt(norm) || 1;
  for (let i = 0; i < vector.length; i += 1) {
    vector[i] = vector[i] / norm;
  }

  return vector;
}

module.exports = {
  embedText,
};

File: search_engine_rust/src/embedding.rs
use crate::utils::normalize_text;

pub fn fnv1a_hash_bytes(bytes: &[u8]) -> u32 {
    let mut hash: u32 = 2166136261;
    for b in bytes {
        hash ^= *b as u32;
        hash = hash.wrapping_mul(16777619);
    }
    hash
}

pub fn embed_text(text: &str, dims: usize, ngram_min: usize, ngram_max: usize) -> Vec<f32> {
    let normalized = normalize_text(text);
    let padded = format!(" {} ", normalized);
    let mut vec = vec![0.0f32; dims];

    let bytes = padded.as_bytes();
    for n in ngram_min..=ngram_max {
        if bytes.len() < n { continue; }
        for i in 0..=bytes.len() - n {
            let gram = &bytes[i..i + n];
            let idx = (fnv1a_hash_bytes(gram) as usize) % dims;
            vec[idx] += 1.0;
        }
    }

    let mut norm = 0.0f32;
    for v in &vec {
        norm += v * v;
    }
    norm = norm.sqrt();
    if norm > 0.0 {
        for v in &mut vec {
            *v /= norm;
        }
    }
    vec
}
